src/test/java/org/broadinstitute/hellbender/tools/recalibration/ContextCovariateUnitTest.java

```diff
@@ -29,38 +35,44 @@ public void initCache() {
-
-        verifyCovariateArray(readCovariates.getMismatchesKeySet(), RAC.MISMATCHES_CONTEXT_SIZE, clippedRead, covariate);
-        verifyCovariateArray(readCovariates.getInsertionsKeySet(), RAC.INDELS_CONTEXT_SIZE, clippedRead, covariate);
-        verifyCovariateArray(readCovariates.getDeletionsKeySet(),  RAC.INDELS_CONTEXT_SIZE,  clippedRead, covariate);
-    }
+        final Random rnd = Utils.getRandomGenerator();
 
-    public static void verifyCovariateArray(int[][] values, int contextSize, SAMRecord read, Covariate contextCovariate) {
-        for (int i = 0; i < values.length; i++)
-            Assert.assertEquals(contextCovariate.formatKey(values[i][0]), expectedContext(read, i, contextSize));
+        for(int i = 0; i < 10; i++) {
+            final SAMRecord read = ArtificialSAMUtils.createRandomRead(1000);
+            read.setReadNegativeStrandFlag(rnd.nextBoolean());
+            final SAMRecord clippedRead = ReadClipper.clipLowQualEnds(read, RAC.LOW_QUAL_TAIL, ClippingRepresentation.WRITE_NS);
+            final ReadCovariates readCovariates = new ReadCovariates(read.getReadLength(), 1);
+            covariate.recordValues(read, readCovariates);
 
+            verifyCovariateArray(readCovariates.getMismatchesKeySet(), RAC.MISMATCHES_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+            verifyCovariateArray(readCovariates.getInsertionsKeySet(), RAC.INDELS_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+            verifyCovariateArray(readCovariates.getDeletionsKeySet(), RAC.INDELS_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+        }
     }
 
-    public static String expectedContext (SAMRecord read, int offset, int contextSize) {
-        final String bases = stringFrom(read.getReadBases());
-        String expectedContext = null;
-        if (offset - contextSize + 1 >= 0) {
-            String context = bases.substring(offset - contextSize + 1, offset + 1);
-            if (!context.contains("N"))
-                expectedContext = context;
+    public static void verifyCovariateArray(int[][] values, int contextSize, SAMRecord read, Covariate contextCovariate, final byte lowQualTail) {
+        for (int i = 0; i < values.length; i++) {
+            Assert.assertEquals(contextCovariate.formatKey(values[i][0]), expectedContext(read, i, contextSize, lowQualTail), "offset " + i);
         }
-        return expectedContext;
     }
 
-    private static String stringFrom(byte[] array) {
-        String s = "";
-        for (byte value : array)
-            s += (char) value;
-        return s;
+    public static String expectedContext (final SAMRecord originalRead, final int offset, final int contextSize, final byte lowQualTail) {
+        final byte[] strandedBaseArray = getStrandedBytes(originalRead, lowQualTail);
+        final int strandedOffset = getStrandedOffset(originalRead.getReadNegativeStrandFlag(), offset, strandedBaseArray.length);
```

Comment:
So now you're calling into these utility methods in the covariate to compute expected values for your tests. But what is testing the correctness of these utility methods themselves (`getStrandedBytes()` and `getStrandedOffset()`)? Shouldn't there be unit tests that tests those methods individually?